### DISCUSSION TITLE
Question on Castorini UniCOiL model in HF (not sure where to ask.)

### DISCUSSION THREAD

**Discussion by PrithivirajDamodaran** [castorini/anserini#2419] [General]
Hello Team, 

I am looking at UniCOiL paper and the models in HuggingFace. I understand vanilla UniCOiL does just term weighting without any expansions and usually expansion models are added post-hoc. But there are `castorini/unicoil-msmarco-passage` and `castorini/unicoil-noexp-msmarco-passage` uploaded. Does that mean the former model has some form of expansion baked into it ? I could not figure out the difference between these two from seeing the config files. Please advice.

Also I didn't know what is the right place to ask this, hence here.

Thanks

**Comment by MXueguang** [comment]
Hi @PrithivirajDamodaran, 
You are correct. `castorini/unicoil-msmarco-passage` takes text input together with document expansions (predicted queries).
`castorini/unicoil-noexp-msmarco-passage` is the vanilla uniCOIL without any expansion.

**Reply by lintool** [reply]
To be clear, with `castorini/unicoil-msmarco-passage` you have to expand the corpus first _separately_.

**Reply by PrithivirajDamodaran** [reply]
Thanks for the responses: 

1.) So in the paper `2g` denotes `castorini/unicoil-noexp-msmarco-passage` - am i right ? If so which entry denotes `castorini/unicoil-msmarco-passage`  ?

<img width="886" alt="Screenshot 2024-03-26 at 12 57 11 PM" src="https://github.com/castorini/anserini/assets/7071019/40930216-1679-4a5b-825d-6d65fdccb5ce">




> To be clear, with castorini/unicoil-msmarco-passage you have to expand the corpus first separately.

2.) Do you mean you used `castorini/unicoil-msmarco-passage ` for document / passage expansion for indexing and `castorini/unicoil-noexp-msmarco-passage` for query expansion at query time ? Sorry, I am not fully clear here.

**Reply by MXueguang** [reply]
1) Yes, 2g denotes `castorini/unicoil-noexp-msmarco-passage` and 2h denotes `castorini/unicoil-msmarco-passage`
2. Both `castorini/unicoil-noexp-msmarco-passage` and `castorini/unicoil-msmarco-passage` are encoding model, that assign token weights to each input tokens. They are not doing any expansion.
To get predicted query for expanding document, need to use the `query2doc-T5` model to generate. We have generated queries stored here `https://huggingface.co/datasets/castorini/msmarco_v1_passage_doc2query-t5_expansions`.

To be clarify, an end2end uniCOIL(**with** expansion) pipeline is:
1. Augment the msmarco corpus with predicted queries. 
- The predicted queries are generated by `query2doc-T5`.
2. Concatenate the original passage and the tokens in predicted queries as augmented passage.
3. Encode augmented passage with `castorini/unicoil-msmarco-passage` to get sparse vector embedding
4. Index passage sparse vectors with pyserini Impact Indexer.
5. Encode queries with `castorini/unicoil-msmarco-passage`
6. Conduct search over the index.

On the other hand, an end2end uniCOIL(**without** expansion) pipeline is:
1. Encode msmarco passage with `castorini/unicoil-noexp-msmarco-passage` to get sparse vector embedding
4. Index passage sparse vectors with pyserini Impact Indexer.
5. Encode queries with `castorini/unicoil-noexp-msmarco-passage`
6. Conduct search over the index.

**Reply by PrithivirajDamodaran** [reply]
@MXueguang Thank you for the detailed response, I am independently reproducing ["Improved Learned Sparse Retrieval with
Corpus-Specific Vocabularies" ](https://arxiv.org/pdf/2401.06703.pdf) which is improving on uniCOIL. So I am trying to be informed before I start.
